Qcom/Library/FdtTargetMmcSdhciLib/QcomTargetMmcSdhciLib.c

```diff
@@ -100,39 +104,120 @@ VOID LibQcomTargetMmcSdhciInit(INIT_SLOT_CB InitSlot)
 
     // get regs
     Status = FdtClient->GetNodeProperty (FdtClient, Node, "reg", (CONST VOID **)&FdtU32, &PropertySize);
-    if (EFI_ERROR (Status) || PropertySize != sizeof(*FdtU32)*4) {
+    if (EFI_ERROR (Status)) {
+      DEBUG((DEBUG_INFO, "FDT read error: \"reg\"\n"));
+      continue;
+    }
+    // get reg names
+    Status = FdtClient->GetNodeProperty(FdtClient, Node, "reg-names", (CONST VOID **)&FdtChar8, NULL);
+    if (EFI_ERROR(Status)) {
+      DEBUG((DEBUG_INFO, "FDT read error: \"reg-names\"\n"));
       continue;
     }
 
     // get base addresses
-    config.sdhc_base = SwapBytes32 (FdtU32[0]);
-    config.pwrctl_base = SwapBytes32 (FdtU32[2]);
+    Index = 0;
+    for (TmpChar8 = FdtChar8; *TmpChar8; TmpChar8 += 1 + AsciiStrLen(TmpChar8)) {
+      if (AsciiStrnCmp("hc_mem", TmpChar8, 6) == 0) {
+        config.sdhc_base = SwapBytes32(FdtU32[Index * 2]);
+      }
+      else if (AsciiStrnCmp("core_mem", TmpChar8, 8) == 0) {
+        config.pwrctl_base = SwapBytes32(FdtU32[Index * 2]);
+      }
+      else {
+        DEBUG((DEBUG_INFO, "Unhandled register: \"%a\" = 0x%08X\n", TmpChar8, SwapBytes32(FdtU32[Index * 2])));
+      }
+      //not handled: "cmdq_mem"
+      Index++;
+    }
+
     DEBUG ((DEBUG_INFO, "sdhc_base: %x\n", config.sdhc_base));
     DEBUG ((DEBUG_INFO, "pwrctl_base: %x\n", config.pwrctl_base));
 
+    // find interrupt parent
+    for (TmpNode = Node; !FdtClient->HasNodeProperty(FdtClient, TmpNode, "interrupt-parent"); Status = FdtClient->GetParentNode(FdtClient, TmpNode, &TmpNode)) {
```

Comment:
Is that loop needed? I think that linux code only checks the current node and the root node.
Correct me if I'm wrong.